miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppCustomPermissionCache.kt

```diff
@@ -0,0 +1,60 @@
+package com.rakuten.tech.mobile.miniapp.js
+
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.annotation.VisibleForTesting
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+
+/**
+ * A class to read and store the grant results of custom permissions per MiniApp.
+ */
+internal class MiniAppCustomPermissionCache(context: Context) {
+    @VisibleForTesting
+    val prefs: SharedPreferences = context.getSharedPreferences(
+        "com.rakuten.tech.mobile.miniapp.custom.permissions.cache", Context.MODE_PRIVATE
+    )
+
+    /**
+     * Reads the grant results from SharedPreferences.
+     * @param [miniAppId] the key provided to find the stored results per MiniApp
+     * @return [MiniAppCustomPermission] an object to hold the results per MiniApp.
+     */
+    fun readPermissions(miniAppId: String): MiniAppCustomPermission {
+        if (prefs.contains(miniAppId)) {
+            return Gson().fromJson(
```

Comment:
Same thing here as on `storePermissions`: It's possible that they have called `storePermission` with just _some_ of the permissions instead of all of them. So maybe we should combine the cached list with the `defaultDeniedList` before returning?

OR we could fix this in the `storePermissions` function by combining the supplied list of permissions with `defaultDeniedList` before saving to the cache?

Reply:
Thanks, I think fixing `storePermissions` is okay.